doc/neps/ufunc-overrides.rst

```diff
@@ -106,163 +116,193 @@ Returning ``NotImplemented`` to user should not happen. Moreover::
   immediately.
 
-If none of the input arguments has a ``__numpy_ufunc__`` method, the
+If none of the input arguments has an ``__array_ufunc__`` method, the
 execution falls back on the default ufunc behaviour.
 
+Subclass hierarchies
+--------------------
+
+Hierarchies of such containers (say, a masked quantity), are most easily
+constructed if methods consistently use :func:`super` to pass through
+the class hierarchy [7]_.  To support this, :class:`ndarray` has its own
+``__array_ufunc__`` method (which is equivalent to ``getattr(ufunc,
+method)(*inputs, **kwargs)``, i.e., if any of the (adjusted) inputs
+still defines ``__array_ufunc__`` that will be called in turn). This
+should be particularly useful for container-like subclasses of
+:class:`ndarray`, which add an attribute like a unit or mask to a
+regular :class:`ndarray`. Such classes can do possible adjustment of the
+arguments relevant to their own class, pass on to another class in the
+hierarchy using :func:`super` until the Ufunc is actually done, and then
+do possible adjustments of the outputs.
+
+Turning Ufuncs off
+------------------
+
+For some classes, Ufuncs make no sense, and, like for other special
+methods [8]_, one can indicate Ufuncs are not available by setting
+``__array_ufunc__`` to :obj:`None`.  Inside a Ufunc, this is
+equivalent to unconditionally return :obj:`NotImplemented`, and thus
+will lead to a :exc:`TypeError` (unless another operand implements
+``__array_ufunc__`` and knows how to deal with the class).
+
+.. [7] https://rhettinger.wordpress.com/2011/05/26/super-considered-super/
+
+.. [8] https://docs.python.org/3/reference/datamodel.html#specialnames
 
 In combination with Python's binary operations
 ----------------------------------------------
 
-The ``__numpy_ufunc__`` mechanism is fully independent of Python's
+The ``__array_ufunc__`` mechanism is fully independent of Python's
 standard operator override mechanism, and the two do not interact
 directly.
 
-They however have indirect interactions, because NumPy's ``ndarray``
-type implements its binary operations via Ufuncs. Effectively, we have::
-
-    class ndarray(object):
+They have indirect interactions, however, because NumPy's
+:class:`ndarray` type implements its binary operations via Ufuncs.  For
+most numerical classes, the easiest way to override binary operations is
+thus to define ``__array_ufunc__`` and override the corresponding
+Ufunc. The class can then, like :class:`ndarray` itself, define the
+binary operators in terms of Ufuncs. Here, one has to take some care.
+E.g., the simplest implementation would be::
+
+    class ArrayLike(object):
+        def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+            ...
+            return result
         ...
         def __mul__(self, other):
-            return np.multiply(self, other)
+            return self.__array_ufunc__(np.multiply, '__call__', self, other)
```

Comment:
We should also mention the other obvious implementation, which calls the ufunc directly instead of `__array_ufunc__`:
```python
class ArrayLike(object):
    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
        ...
    def __mul__(self, other):
        return np.multiply(self, other)
```

This has similar issues handling "opt-out" classes, but there's also a fix:
```python
# this lookup table should probably be distributed with numpy
OPERATOR_LOOKUP = {np.multiply: operator.multiply, ...}

class ArrayLike(object):
    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
        func = getattr(ufunc, method)
        if method == '__call__' and 'out' not in kwargs:
            func = OPERATOR_LOOKUP.get(ufunc, func)
        # avoid infinite recursion
        inputs = [np.asarray(x) if x is self else x for x in inputs]
        return func(*inputs)
```

I like this approach better because it doesn't involve calling private methods, which entails the need to reimplement some of the rules of `__array_ufunc__` in your own methods. But I think both can be made to work.

From an API perspective, the difference is that `np.multiply(array_like, opt_out)` is well defined (equivalent to `array_like * opt_out`) instead of raising a TypeError, even though `opt_out.__array_ufunc__ is None`. But I'm actually OK with either option (even for `numpy.ndarray`): things only get terribly complex if it's possible for `np.multiply` and `*` to do *different* things.